src/synapse/api.py

```diff
@@ -0,0 +1,459 @@
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+class NetworkError(APIError):
+    """Exception raised when requesting API fails due network issues."""
+
+
+class GetNonceError(APIError):
+    """Exception raised when getting nonce fails."""
+
+
+class GetVersionError(APIError):
+    """Exception raised when getting version fails."""
+
+
+class VersionUnexpectedContentError(GetVersionError):
+    """Exception raised when output of getting version is unexpected."""
+
+
+class GetRoomIDError(APIError):
+    """Exception raised when getting room id fails."""
+
+
+class GetUserIDError(APIError):
+    """Exception raised when getting user id fails."""
+
+
+class UserExistsError(APIError):
+    """Exception raised when checking if user exists fails."""
+
+
+class GetAccessTokenError(APIError):
+    """Exception raised when getting access token fails."""
+
+
+class RegisterUserError(APIError):
+    """Exception raised when registering user fails."""
+
+
+# admin_access_token is not a password
+def register_user(
+    registration_shared_secret: str,
+    user: User,
+    server: typing.Optional[str] = None,
+    admin_access_token: typing.Optional[str] = None,  # nosec
+) -> str:
+    """Register user.
+
+    Args:
+        registration_shared_secret: secret to be used to register the user.
+        user: user to be registered.
+        server: to be used to create the user id.
+        admin_access_token: admin access token to get user's access token if it exists.
+
+    Raises:
+        RegisterUserError: if there was an error registering the user.
+
+    Returns:
+        Access token to be used by the user.
+    """
+    # get nonce
+    nonce = _get_nonce()
+    # generate mac
+    hex_mac = _generate_mac(
+        shared_secret=registration_shared_secret,
+        nonce=nonce,
+        user=user.username,
+        password=user.password,
+        admin=user.admin,
+    )
+    # build data
+    data = {
+        "nonce": nonce,
+        "username": user.username,
+        "password": user.password,
+        "mac": hex_mac,
+        "admin": user.admin,
+    }
+    # finally register user
+    try:
+        res = _do_request("POST", REGISTER_URL, json=data)
+        return res.json()["access_token"]
+    except UserExistsError as exc:
+        logger.warning("User %s exists, getting the access token.", user.username)
+        if not server or not admin_access_token:
+            raise RegisterUserError(
+                f"User {user.username} exists but there is no server/admin access token set."
+            ) from exc
+        return get_access_token(
+            user=user, server=str(server), admin_access_token=str(admin_access_token)
+        )
+    except (requests.exceptions.JSONDecodeError, TypeError, KeyError) as exc:
+        logger.exception("Failed to decode access_token: %r. Received: %s", exc, res.text)
+        raise RegisterUserError(str(exc)) from exc
```

Comment:
Similar comment as before, because `from exc` is used here, the error message in that exception will already be shown in the stack trace so the `RegisterUserError` could be passed a message which has a bit more context about what this module is trying to do

Reply:
Related: ISD-1070